js/background.js

```diff
@@ -233,14 +233,13 @@
     }
 
     function onSentMessage(ev) {
-        var now = new Date().getTime();
         var data = ev.data;
 
         var message = new Whisper.Message({
             source         : textsecure.storage.user.getNumber(),
             sourceDevice   : data.device,
             sent_at        : data.timestamp,
-            received_at    : now,
+            received_at    : data.receivedAt || Date.now(),
```

Comment:
No, `data.receivedAt` is not when it was sent, but when we get it. We set `envelope.receivedAt` right as we receive it from the websocket, and then that value is flowed here. It is also flowed here during retries as you mention.

A message's `received_at`, as you guessed, is very similar to your `dateForSorting` because it is indeed what we sort on first. I recently updated it so we sort on `received_at` and then `sent_at` for disambiguation.